core/src/main/java/com/linecorp/armeria/client/endpoint/WeightRampingUpStrategy.java

```diff
@@ -0,0 +1,470 @@
+            return endpointSelector;
+        }
+
+        // Only executed by the executor.
+        private void updateEndpoints(List<Endpoint> newEndpoints) {
+            unhandledNewEndpoints = null;
+            if (rampingUpTaskWindowNanos > 0) {
+                // Check whether we can ramp up with the previous ramped up endpoints which are at the last
+                // of the rampingUpEndpointsEntries.
+                if (shouldRampUpWithPreviousRampedUpEntry()) {
+                    final Set<EndpointAndStep> newlyAddedEndpoints = filterOldEndpoints(newEndpoints);
+                    if (!newlyAddedEndpoints.isEmpty()) {
+                        updateWeightAndStep(newlyAddedEndpoints);
+                        endpointsRampingUp.getLast().addEndpoints(newlyAddedEndpoints);
+                    }
+                    buildEndpointSelector();
+                    return;
+                }
+
+                // Check whether we can ramp up with the next scheduled endpointsRampingUpEntry.
+                if (shouldRampUpWithNextScheduledEntry()) {
+                    // unhandledNewEndpoints will be ramped up when updateWeightAndStep() is executed.
+                    unhandledNewEndpoints = newEndpoints;
+                    return;
+                }
+            }
+
+            final Set<EndpointAndStep> newlyAddedEndpoints = filterOldEndpoints(newEndpoints);
+            if (newlyAddedEndpoints.isEmpty()) {
+                // newlyAddedEndpoints is empty which means that endpointsFinishedRampingUp are changed.
+                // So rebuild the endpoint selector.
+                buildEndpointSelector();
+                return;
+            }
+
+            updateWeightAndStep(newlyAddedEndpoints);
+
+            // Check again because newlyAddedEndpoints can be removed in the updateWeightAndStep method.
+            if (!newlyAddedEndpoints.isEmpty()) {
+                final ScheduledFuture<?> scheduledFuture = executor.scheduleAtFixedRate(
+                        this::updateWeightAndStep, rampingUpIntervalMillis,
+                        rampingUpIntervalMillis, TimeUnit.MILLISECONDS);
+                final EndpointsRampingUpEntry entry = new EndpointsRampingUpEntry(
+                        newlyAddedEndpoints, scheduledFuture, ticker, rampingUpIntervalMillis);
+                endpointsRampingUp.add(entry);
+            }
+            buildEndpointSelector();
+        }
+
+        private void buildEndpointSelector() {
+            final ImmutableList.Builder<Endpoint> targetEndpointsBuilder = ImmutableList.builder();
+            targetEndpointsBuilder.addAll(endpointsFinishedRampingUp);
+            endpointsRampingUp.forEach(
+                    entry -> entry.endpointAndSteps().forEach(
+                            endpointAndStep -> targetEndpointsBuilder.add(
+                                    endpointAndStep.endpoint().withWeight(endpointAndStep.currentWeight()))));
```

Comment:
nit: How about using stream API or loop to avoid nested indentations?
```java
endpointsRampingUp.stream()
                  .flatMap(entry -> entry.endpointAndSteps().stream())
                  .map(endpointAndStep -> endpointAndStep.endpoint().withWeight(endpointAndStep.currentWeight()))
                  .forEach(targetEndpointsBuilder::add);
// or
for (EndpointsRampingUpEntry entry : endpointsRampingUp) {
    for (EndpointAndStep endpointAndStep : entry.endpointAndSteps()) {
        targetEndpointsBuilder.add(
                endpointAndStep.endpoint().withWeight(endpointAndStep.currentWeight()));
    }
}
```

Reply:
Let me use the foreach version. 😄